juju/juju
# [JUJU-1024] Retry with a different instance on Azure quota errors

When starting an Azure instance, there can be a quota error. It's possible to retry with a different instance type and have success. We try the lowest cost instance type first - if that results in a quota error, we throw that one away and try again with the next most expensive instance type.

This was seen in QA where the credential used was not able to start a c core instance in some of the D(s) family.

We don't currently attempt to parse the rather long error info to try and interpret the specific reason for the quota error; for now, it's enough to retry a few times with the bad instance types omitted.

Also, the regexp used to parse the instance names is improved so that the cost array only needs to contain the actual name part.

## QA steps

Using the juju QA credential, bootstrap with 2 cores

`juju bootstrap azure --show-log --constraints="cores=2 mem=6144M root-disk=30720M"`

Observe quota warnings
```
14:02:07 INFO  juju.environs.instances image.go:107 find instance - using amd64 image of type Standard_D2as_v5 with id: Canonical:0001-com-ubuntu-server-focal:20_04-lts:latest
14:02:07 INFO  juju.provider.azure environ.go:600 picked agent binaries "2.9.30.1-ubuntu-amd64"
14:02:11 WARN  juju.provider.azure environ.go:567 Standard_D2as_v5 quota exceeded error: "Operation could not be completed as it results in exceeding approved standardDASv5Family Cores quota. Additional details - Deployment Model: Resource Manager, Location: centralus, Current Limit: 0, Current Usage: 0, Additional Required: 2, (Minimum) New Limit Required: 2. Submit a request for Quota increase at https://aka.ms/ProdportalCRP/#blade/Microsoft_Azure_Capacity/UsageAndQuota.ReactView/Parameters/%7B%22subscriptionId%22:%222eebf55a-1e02-45d8-a299-02aed8aea00b%22,%22command%22:%22openQuotaApprovalBlade%22,%22quotas%22:[%7B%22location%22:%22centralus%22,%22providerId%22:%22Microsoft.Compute%22,%22resourceName%22:%22standardDASv5Family%22,%22quotaRequest%22:%7B%22properties%22:%7B%22limit%22:2,%22unit%22:%22Count%22,%22name%22:%7B%22value%22:%22standardDASv5Family%22%7D%7D%7D%7D]%7D by specifying parameters listed in the ‘Details’ section for deployment to succeed. Please read more about quota limits at https://docs.microsoft.com/en-us/azure/azure-supportability/per-vm-quota-requests"
```

but the bootstrap will find a good instance and succeed.
